### PR TITLE
fix: Update power CTS to handle new GPU/Memory domains

### PR DESCRIPTION
**Pull request by shubskmr**
Related-To: NEO-13195

### CODE REVIEW

File: utils/test_harness/sysman/src/test_harness_sysman_power.cpp [hunk 1]
```diff
@@ -44,29 +44,43 @@ zes_power_properties_t get_power_properties(zes_pwr_handle_t pPowerHandle) {
...
 }
-uint32_t get_power_limit_count(zes_pwr_handle_t hPower) {
-  uint32_t pCount = 0;
-  EXPECT_EQ(ZE_RESULT_SUCCESS, zesPowerGetLimitsExt(hPower, &pCount, nullptr));
-  EXPECT_GE(pCount, 0);
-  return pCount;
+ze_result_t get_power_limit_count(zes_pwr_handle_t hPower, uint32_t *pCount) {
+  *pCount = 0;
+  ze_result_t status = zesPowerGetLimitsExt(hPower, pCount, nullptr);
+  if (status != ZE_RESULT_ERROR_UNSUPPORTED_FEATURE) {
```
Comment (joshuaranjan) [inline utils/test_harness/sysman/src/test_harness_sysman_power.cpp:50]: Can we return the status directly and move these checks to the caller ?
Hiding this detail under the harness looks easy to miss for the test writer

File: utils/test_harness/sysman/src/test_harness_sysman_power.cpp [hunk 2]
```diff
@@ -85,30 +99,39 @@ void compare_power_descriptor_structures(
...
-void set_power_limits(zes_pwr_handle_t pPowerHandle,
+ze_result_t set_power_limits(zes_pwr_handle_t pPowerHandle,
                       zes_power_sustained_limit_t *pSustained,
                       zes_power_burst_limit_t *pBurst,
                       zes_power_peak_limit_t *pPeak) {
-  EXPECT_EQ(ZE_RESULT_SUCCESS,
-            zesPowerSetLimits(pPowerHandle, pSustained, pBurst, pPeak));
+  ze_result_t status =
+      zesPowerSetLimits(pPowerHandle, pSustained, pBurst, pPeak);
+  if (status != ZE_RESULT_ERROR_UNSUPPORTED_FEATURE) {
```
Comment (joshuaranjan) [inline utils/test_harness/sysman/src/test_harness_sysman_power.cpp:119]: Please handle other errors (other than SUCCESS or UNSUPPORTED)

File: utils/test_harness/sysman/src/test_harness_sysman_power.cpp (hunk 1 above)
Comment (shubskmr) [reply utils/test_harness/sysman/src/test_harness_sysman_power.cpp:50]: `get_power_limit_count` is called by another harness method `get_power_limits_ext`. It is not called directly by the TESTS

File: utils/test_harness/sysman/src/test_harness_sysman_power.cpp (hunk 2 above)
Comment (shubskmr) [reply utils/test_harness/sysman/src/test_harness_sysman_power.cpp:119]: I did not get the request. We are not handling just bypassing the current check `EXPECT_EQ(ZE_RESULT_SUCCESS, status);` when the status is `ZE_RESULT_ERROR_UNSUPPORTED_FEATURE`. This is done since `ZE_RESULT_ERROR_UNSUPPORTED_FEATURE` is the only acceptable case when we should not error out.

Comment (joshuaranjan) [reply utils/test_harness/sysman/src/test_harness_sysman_power.cpp:119]: What happens if error other than UNSUPPORTED is encountered

File: utils/test_harness/sysman/src/test_harness_sysman_power.cpp (hunk 1 above)
Comment (joshuaranjan) [reply utils/test_harness/sysman/src/test_harness_sysman_power.cpp:50]: > `get_power_limit_count` is called by another harness method `get_power_limits_ext`. It is not called directly by the TESTS

Hmm .. That is still risky.
Please create another harness

File: utils/test_harness/sysman/include/test_harness_sysman_power.hpp [hunk 3]
```diff
@@ -17,28 +17,28 @@ std::vector<zes_pwr_handle_t> get_power_handles(zes_device_handle_t device,
...
-void set_power_limits_ext(zes_pwr_handle_t hPower, uint32_t *pCount,
+ze_result_t get_power_limits_ext(zes_pwr_handle_t hPower, uint32_t *pCount,
+    std::vector<zes_power_limit_ext_desc_t> &pPowerLimitsDescriptors);
+ze_result_t set_power_limits_ext(zes_pwr_handle_t hPower, uint32_t *pCount,
                           zes_power_limit_ext_desc_t *pSustained);
 void compare_power_descriptor_structures(
     zes_power_limit_ext_desc_t firstDescriptor,
     zes_power_limit_ext_desc_t secondDescriptor);
-uint32_t get_power_limit_count(zes_pwr_handle_t hPower);
+ze_result_t get_power_limit_count(zes_pwr_handle_t hPower, uint32_t *pCount);
```
Comment (AshwinKumarKulkarni) [inline utils/test_harness/sysman/include/test_harness_sysman_power.hpp:41]: better to add a new overloaded function than modifying existing interface?, thoughts?

File: utils/test_harness/sysman/src/test_harness_sysman_power.cpp (hunk 1 above)
Comment (shubskmr) [reply utils/test_harness/sysman/src/test_harness_sysman_power.cpp]: removed these checks from harness and added in actual TEST

File: utils/test_harness/sysman/include/test_harness_sysman_power.hpp (hunk 3 above)
Comment (shubskmr) [reply utils/test_harness/sysman/include/test_harness_sysman_power.hpp:41]: only power module is calling these power specific harness methods. If we overload original methods will not be used anymore anywhere.

File: utils/test_harness/sysman/src/test_harness_sysman_power.cpp (hunk 2 above)
Comment (shubskmr) [reply utils/test_harness/sysman/src/test_harness_sysman_power.cpp]: removed these from harness

File: conformance_tests/sysman/test_sysman_events/src/test_sysman_events.cpp [hunk 4]
```diff
@@ -453,7 +453,14 @@ TEST_F(
              << _ze_result_t(ZE_RESULT_ERROR_UNSUPPORTED_FEATURE);
     }
     for (auto power_handle : power_handles) {
-      auto energy_threshold = lzt::get_power_energy_threshold(power_handle);
+      zes_energy_threshold_t energy_threshold = {};
+      auto status =
+          lzt::get_power_energy_threshold(power_handle, &energy_threshold);
+      if (status == ZE_RESULT_ERROR_UNSUPPORTED_FEATURE) {
+        continue;
+      } else {
```
Comment (joshuaranjan) [inline conformance_tests/sysman/test_sysman_events/src/test_sysman_events.cpp:461]: minor : else is not needed, since we are continuing

File: conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp [hunk 5]
```diff
@@ -194,14 +199,19 @@ TEST_F(
...
+      if (status == ZE_RESULT_ERROR_UNSUPPORTED_FEATURE) {
+        continue;
+      } else {
+        EXPECT_EQ(status, ZE_RESULT_SUCCESS);
+      }
       zes_power_sustained_limit_t pSustainedLater = {};
       zes_power_burst_limit_t pBurstLater = {};
       zes_power_peak_limit_t pPeakLater = {};
-      lzt::get_power_limits(p_power_handle, &pSustainedLater, &pBurstLater,
+      status = lzt::get_power_limits(p_power_handle, &pSustainedLater, &pBurstLater,
```
Comment (joshuaranjan) [inline conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp:212]: no check for UNSUPPORTED feature here ?

File: conformance_tests/sysman/test_sysman_events/src/test_sysman_events.cpp (hunk 4 above)
Comment (shubskmr) [reply conformance_tests/sysman/test_sysman_events/src/test_sysman_events.cpp:461]: We want to fail on any other return codes other than UNSUPPORTED, hence checking if its equal to SUCCESS.

File: conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp (hunk 5 above)
Comment (shubskmr) [reply conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp:212]: if its UNSUPPORTED for a power handle we will skip based on the first call itself on line 205

File: conformance_tests/sysman/test_sysman_events/src/test_sysman_events.cpp (hunk 4 above)
Comment (joshuaranjan) [reply conformance_tests/sysman/test_sysman_events/src/test_sysman_events.cpp:461]: > We want to fail on any other return codes other than UNSUPPORTED, hence checking if its equal to SUCCESS.

Checking for success is needed.
The "else" is not needed and the code can fall through

File: conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp (hunk 5 above)
Comment (shubskmr) [reply conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp:210]: added everywhere now to be consistent

**Comment by AshwinKumarKulkarni** [comment]
lgtm!!

File: conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp [hunk 6]
```diff
@@ -860,21 +1011,36 @@ TEST_F(
...
+          status = lzt::set_power_limits_ext(power_card_handle, &count_power,
                                     &power_sustained_Max);
-          auto power_limits_descriptors_getMax =
-              lzt::get_power_limits_ext(power_card_handle, &count_power);
+          if (status == ZE_RESULT_ERROR_UNSUPPORTED_FEATURE) {
+            continue;
+          }
+          EXPECT_EQ(status, ZE_RESULT_SUCCESS);
+          std::vector<zes_power_limit_ext_desc_t> power_limits_descriptors_getMax;
+          status = lzt::get_power_limits_ext(power_card_handle, &count_power,
```
Comment (joshuaranjan) [inline conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp:1021]: May be we need the UNSUPPORTED_FEATURE check here

File: conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp [hunk 7]
```diff
@@ -979,7 +1169,7 @@ TEST_F(
                 zesPowerGetProperties(p_power_handle, &pProperties));
 
       EXPECT_GT(pExtProperties.domain, ZES_POWER_DOMAIN_UNKNOWN);
-      EXPECT_LE(pExtProperties.domain, ZES_POWER_DOMAIN_STACK);
+      EXPECT_LE(pExtProperties.domain, ZES_POWER_DOMAIN_GPU);
```
Comment (joshuaranjan) [inline conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp:1172]: Why this change ?

File: conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp (hunk 6 above)
Comment (shubskmr) [reply conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp:1021]: thanks for the catch added now

File: conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp (hunk 7 above)
Comment (shubskmr) [reply conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp:1175]: Its the order in which the enum is defined. GPU can get enumerated now, like in BMG so we should check the value to be lesser than GPU

```
typedef enum _zes_power_domain_t
{
    ZES_POWER_DOMAIN_UNKNOWN = 0,                                           ///< The PUnit power domain level cannot be determined.
    ZES_POWER_DOMAIN_CARD = 1,                                              ///< The PUnit power domain is a card-level power domain.
    ZES_POWER_DOMAIN_PACKAGE = 2,                                           ///< The PUnit power domain is a package-level power domain.
    ZES_POWER_DOMAIN_STACK = 3,                                             ///< The PUnit power domain is a stack-level power domain.
    ZES_POWER_DOMAIN_MEMORY = 4,                                            ///< The PUnit power domain is a memory-level power domain.
    ZES_POWER_DOMAIN_GPU = 5,                                               ///< The PUnit power domain is a GPU-level power domain.
    ZES_POWER_DOMAIN_FORCE_UINT32 = 0x7fffffff

} zes_power_domain_t;
```

Comment (joshuaranjan) [reply conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp:1175]: I am not sure why only till STACK was used previously, if the spec had references till GPU

We should be using EXPECT_LT (ZES_POWER_DOMAIN_FORCE_UINT32)

Comment (shubskmr) [reply conformance_tests/sysman/test_sysman_power/src/test_sysman_power.cpp]: updated this

**Review by joshuaranjan — APPROVED** [review]
THanks for the updates!
LGTM.
@AshwinKumarKulkarni  please also check

**Review by AshwinKumarKulkarni — APPROVED** [review]
